crates/red_knot_python_semantic/src/types.rs

```diff
@@ -195,6 +195,8 @@ pub enum Type<'db> {
     LiteralString,
     /// A bytes literal
     BytesLiteral(BytesLiteralType<'db>),
+    /// A heterogeneous tuple type, with elements of the given types in source order.
+    TupleLiteral(TupleLiteralType<'db>),
```

Comment:
I saw there was some discussion of this already, but I'm not sure I see the rationale for calling this `TupleLiteral` vs just `Tuple`. The current representation in this PR can already handle all tuple types other than `tuple[T, ...]`, and it wouldn't take much to add support for that, too (though I'm not suggesting we do it in this PR, and I'm not sure yet whether we should support that via this variant or via the regular builtin tuple type from typeshed). This representation can definitely already handle more tuple types than just literals.

Unless I'm missing the rationale, I would just call this `Type::Tuple`, and the contained struct `TupleType`.